### ISSUE TITLE
How to make QueryGallery Dataset from Pytorch ImageFolder

### ISSUE THREAD

**Issue by ohernpaul** [catalyst-team/catalyst#935]
#### What is your question?
What is the correct way of making a QueryGallery Dataset from Pytorch ImageFolder Dataset? 

The reason I ask is because the MNIST example isn't necessarily the same. 

I am confused by the mnist test dataset being preshuffled and having the gallery images randomly assigned to the last 20%. It seems there's no guarantee that each class will be assigned a gallery image. Please correct me if my understanding is incorrect. 

#### Code
My attempted class:
`
class HandsQGDataset(QueryGalleryDataset):
    
    def __init__(self, path, transform=None, gallery_fraq=0.2):
        self.path = path
        self.transform = transform
        
        #self.shuf_ds = ShuffledImageFolder(self.path, self.transform)
        self.shuf_ds = ImageFolder(self.path, transform = self.transform)
        
        self.gallery_fraq = gallery_fraq
        
        self._gallery_size = int(self.gallery_fraq * len(self.shuf_ds))
        self._query_size = len(self.shuf_ds) - self._gallery_size
        
        self._is_query = torch.zeros(len(self.shuf_ds)).type(torch.bool)
        self._is_query[: self._query_size] = True
        
        
    def __getitem__(self, idx: int):
        """
        Get item method for dataset

        Args:
            idx: index of the object

        Returns:
            Dict with features, targets and is_query flag
        """
        image, label = self.shuf_ds[idx]
        return {
            "features": image,
            "targets": label,
            "is_query": self._is_query[idx],
        }
    
    def __len__(self) -> int:
        """Length"""
        return len(self.shuf_ds)

    @property
    def gallery_size(self) -> int:
        """Query Gallery dataset should have gallery_size property"""
        return self._gallery_size

    @property
    def query_size(self) -> int:
        """Query Gallery dataset should have query_size property"""
        return self._query_size
        
`

I have also tried to create a custom preshuffled image dataset that is called by HandsQGDataset

`
class ShuffledImageFolder(Dataset):
    
    def getImgsLabs(self):
        t = list(iter(os.walk(self.path, topdown=False)))
        
        out = [[''.join([t[i][0] + '/', t[i][2][0]]),int(t[i][0].split('/')[-1])] \
               for i in range(len(t)) for j in range(len(t[i][2])) \
               if 'ipy' not in t[i][0]]
        
        np.random.shuffle(out)
        
        return out
    
    def loadImg(self, file_path):
        return io.imread(file_path)
    
    def __init__(self, path, transform=None):
        self.path = path
        self.transform = transform
        
        self._ds = self.getImgsLabs()
        
    def __getitem__(self, idx):
        path, lab = self._ds[idx]
        img = self.loadImg(path)
        
        if self.transform is not None:
            img = self.transform(img)
        
        return img, lab
    
    def __len__(self):
        return len(self._ds)
        
`

#### What's your environment?
Catalyst version: 20.8.2
PyTorch version: 1.1.0
TensorFlow version: N/A
TensorBoard version: 2.1.0
OS: Mac OSX 10.15.5
Python version: 3.7
```


### Additional context
The training behavior seems correct and the loss decreases, but in Validation with cmc01 while the dataset has not been shuffled, I continually get a value of 0.0152

iI Validation with cmc01 while the dataset has been shuffled, I continually get a value of 1.00

![image](https://user-images.githubusercontent.com/13112890/92516971-30f55780-f1d3-11ea-887b-7da7ecbdfddc.png)

**Comment by elephantmipt** [comment]
Hi @ohernpaul! Thank you for your question. You are right about `MnistQGDataset` it's randomly assigned to the gallery last 20% of the dataset. The correctness of this partition manly guaranteed by the number of examples in the gallery and a small number of labels. Anyway, MNIST dataset is just a small example that everything works well.
In many academic datasets (omniglot etc) there is fixed query/gallery partition. So you don't need to create partition by yourself. If you are going to create your own query/gallery partition you should guarantee the correctness of this partition by your self (use stratified split or something similar).
Therefore I'm not sure that constant CMC-score is the main problem here. Maybe for your partition cmc 0.0152 is a maximal score you can get, and even random initialized network tend to put similar objects together in embedding space. Also, your loss should be lower than the margin.
In my opinion, you should fix your query/gallery partition and ensure that this partition is correct before you start your experiments.

**Comment by Scitator** [comment]
Looks like answered :)

**Comment by Scitator** [comment]
Nevertheless, if you have any questions - feel free to join our slack https://join.slack.com/t/catalyst-team-core/shared_invite/zt-d9miirnn-z86oKDzFMKlMG4fgFdZafw
